File: src/rule.cc
// Copyright 2015 Google Inc. All rights reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// +build ignore

#include "rule.h"

#include "expr.h"
#include "log.h"
#include "parser.h"
#include "stringprintf.h"
#include "strutil.h"
#include "symtab.h"

Rule::Rule() : is_double_colon(false), is_suffix_rule(false), cmd_lineno(0) {}

void Rule::ParseInputs(const std::string_view& inputs_str) {
  bool is_order_only = false;
  for (auto const& input : WordScanner(inputs_str)) {
    if (input == "|") {
      is_order_only = true;
      continue;
    }
    Symbol input_sym = Intern(TrimLeadingCurdir(input));
    (is_order_only ? order_only_inputs : inputs).push_back(input_sym);
  }
}

void Rule::ParsePrerequisites(const std::string_view& line,
                              size_t separator_pos,
                              const RuleStmt* rule_stmt) {
  // line is either
  //    prerequisites [ ; command ]
  // or
  //    target-prerequisites : prereq-patterns [ ; command ]
  // First, separate command. At this point separator_pos should point to ';'
  // unless null.
  std::string_view prereq_string = line;
  if (separator_pos != std::string::npos &&
      rule_stmt->sep != RuleStmt::SEP_SEMICOLON) {
    CHECK(line[separator_pos] == ';');
    // TODO: Maybe better to avoid Intern here?
    cmds.push_back(Value::NewLiteral(
        Intern(TrimLeftSpace(line.substr(separator_pos + 1))).str()));
    prereq_string = line.substr(0, separator_pos);
  }

  if ((separator_pos = prereq_string.find(':')) == std::string::npos) {
    // Simple prerequisites
    ParseInputs(prereq_string);
    return;
  }

  // Static pattern rule.
  if (!output_patterns.empty()) {
    ERROR_LOC(loc, "*** mixed implicit and normal rules: deprecated syntax");
  }

  // Empty static patterns should not produce rules, but need to eat the
  // commands So return a rule with no outputs nor output_patterns
  if (outputs.empty()) {
    return;
  }

  std::string_view target_prereq = prereq_string.substr(0, separator_pos);
  std::string_view prereq_patterns = prereq_string.substr(separator_pos + 1);

  for (std::string_view target_pattern : WordScanner(target_prereq)) {
    target_pattern = TrimLeadingCurdir(target_pattern);
    for (Symbol target : outputs) {
      if (!Pattern(target_pattern).Match(target.str())) {
        WARN_LOC(loc, "target `%s' doesn't match the target pattern",
                 target.c_str());
      }
    }
    output_patterns.push_back(Intern(target_pattern));
  }

  if (output_patterns.empty()) {
    ERROR_LOC(loc, "*** missing target pattern.");
  }
  if (output_patterns.size() > 1) {
    ERROR_LOC(loc, "*** multiple target patterns.");
  }
  if (!IsPatternRule(output_patterns[0].str())) {
    ERROR_LOC(loc, "*** target pattern contains no '%%'.");
  }
  ParseInputs(prereq_patterns);
}

std::string Rule::DebugString() const {
  std::vector<std::string> v;
  v.push_back(StringPrintf("outputs=[%s]", JoinSymbols(outputs, ",").c_str()));
  v.push_back(StringPrintf("inputs=[%s]", JoinSymbols(inputs, ",").c_str()));
  if (!order_only_inputs.empty()) {
    v.push_back(StringPrintf("order_only_inputs=[%s]",
                             JoinSymbols(order_only_inputs, ",").c_str()));
  }
  if (!output_patterns.empty()) {
    v.push_back(StringPrintf("output_patterns=[%s]",
                             JoinSymbols(output_patterns, ",").c_str()));
  }
  if (is_double_colon)
    v.push_back("is_double_colon");
  if (is_suffix_rule)
    v.push_back("is_suffix_rule");
  if (!cmds.empty()) {
    v.push_back(StringPrintf("cmds=[%s]", JoinValues(cmds, ",").c_str()));
  }
  return JoinStrings(v, " ");
}

File: src-rs/rule.rs
/*
Copyright 2025 Google LLC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

     https://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

use std::fmt::Debug;
use std::sync::Arc;

use anyhow::Result;
use bytes::Bytes;
use memchr::memchr;

use crate::expr::Value;
use crate::loc::Loc;
use crate::stmt::{RuleSep, RuleStmt};
use crate::strutil::{Pattern, trim_leading_curdir, word_scanner};
use crate::symtab::{Symbol, intern};
use crate::{error_loc, warn_loc};

#[derive(Clone)]
pub struct Rule {
    pub outputs: Vec<Symbol>,
    pub inputs: Vec<Symbol>,
    pub order_only_inputs: Vec<Symbol>,
    pub output_patterns: Vec<Symbol>,
    pub validations: Vec<Symbol>,
    pub is_double_colon: bool,
    pub is_suffix_rule: bool,
    pub cmds: Vec<Arc<Value>>,
    pub loc: Loc,
    pub cmd_loc: Option<Loc>,
}

impl Rule {
    pub fn new(loc: Loc, is_double_colon: bool) -> Self {
        Self {
            outputs: Vec::new(),
            inputs: Vec::new(),
            order_only_inputs: Vec::new(),
            output_patterns: Vec::new(),
            validations: Vec::new(),
            is_double_colon,
            is_suffix_rule: false,
            cmds: Vec::new(),
            loc,
            cmd_loc: None,
        }
    }

    fn parse_inputs(&mut self, inputs_str: &Bytes) {
        let mut is_order_only = false;
        for input in word_scanner(inputs_str) {
            if input == b"|" {
                is_order_only = true;
                continue;
            }
            let input_sym = intern(inputs_str.slice_ref(trim_leading_curdir(input)));
            if is_order_only {
                self.order_only_inputs.push(input_sym);
            } else {
                self.inputs.push(input_sym);
            }
        }
    }

    pub fn parse_prerequisites(
        &mut self,
        line: &Bytes,
        separator_pos: Option<usize>,
        rule_stmt: &RuleStmt,
    ) -> Result<()> {
        // line is either
        //    prerequisites [ ; command ]
        // or
        //    target-prerequisites : prereq-patterns [ ; command ]
        // First, separate command. At this point separator_pos should point to ';'
        // unless null.
        let mut prereq_string = line.clone();
        if let Some(separator_pos) = separator_pos
            && rule_stmt.sep != RuleSep::Semicolon
        {
            assert!(line[separator_pos] == b';');
            let value = line.slice(separator_pos + 1..);
            self.cmds.push(Arc::new(Value::Literal(None, value)));
            prereq_string = line.slice(..separator_pos);
        }

        let Some(separator_pos) = memchr(b':', &prereq_string) else {
            // Simple prerequisites
            self.parse_inputs(&prereq_string);
            return Ok(());
        };

        // Static pattern rule.
        if !self.output_patterns.is_empty() {
            error_loc!(
                Some(&self.loc),
                "*** mixed implicit and normal rules: deprecated syntax"
            );
        }

        // Empty static patterns should not produce rules, but need to eat the
        // commands So return a rule with no outputs nor output_patterns
        if self.outputs.is_empty() {
            return Ok(());
        }

        let target_prereq = prereq_string.slice(..separator_pos);
        let prereq_patterns = prereq_string.slice(separator_pos + 1..);

        for target_pattern in word_scanner(&target_prereq) {
            let target_pattern = target_prereq.slice_ref(trim_leading_curdir(target_pattern));
            let pat = Pattern::new(target_pattern.clone());
            for target in &self.outputs {
                if !pat.matches(&target.as_bytes()) {
                    warn_loc!(
                        Some(&self.loc),
                        "target `{target}' doesn't match the target pattern",
                    );
                }
            }
            self.output_patterns.push(intern(target_pattern));
        }

        if self.output_patterns.is_empty() {
            error_loc!(Some(&self.loc), "*** missing target pattern.");
        }
        if self.output_patterns.len() > 1 {
            error_loc!(Some(&self.loc), "*** multiple target patterns.");
        }
        if !is_pattern_rule(&self.output_patterns.first().unwrap().as_bytes()) {
            error_loc!(Some(&self.loc), "*** target pattern contains no '%'.");
        }
        self.parse_inputs(&prereq_patterns);
        Ok(())
    }
}

impl Debug for Rule {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "outputs={:?} inputs={:?}", self.outputs, self.inputs)?;
        if !self.order_only_inputs.is_empty() {
            write!(f, " order_only_inputs={:?}", self.order_only_inputs)?;
        }
        if !self.output_patterns.is_empty() {
            write!(f, " output_patterns={:?}", self.output_patterns)?;
        }
        if self.is_double_colon {
            write!(f, " is_double_colon")?;
        }
        if self.is_suffix_rule {
            write!(f, " is_suffix_rule")?;
        }
        if !self.cmds.is_empty() {
            write!(f, " cmds={:?}", self.cmds)?;
        }
        Ok(())
    }
}

pub fn is_pattern_rule(target: &[u8]) -> bool {
    memchr(b'%', target).is_some()
}
